make it easier to work with metadata by providing more default behaviors

bkamins opened the pull request:
Fixes https://github.com/JuliaData/DataAPI.jl/issues/55
Fixes https://github.com/JuliaData/DataAPI.jl/issues/51

@tokazama - in relation to #51 I ended up recommending to return a dictionary (not just an iterator). Do you think it would be acceptable, or you would prefer to require only an iterator of pairs?

The benefit of dictionary would be that later it is easier for users to work with it. I assume that if a given value does not have already an underlying dictionary to store metadata (most have) then just `Dict` will be used since this is a non-lazy interface.

If someone wants a lazy iterator then using `metadatakeys` and then getting the required metadata is an option that is available.

Tokazama commented:
> Do you think it would be acceptable, or you would prefer to require only an iterator of pairs?

I think the tricky part is where that `style` argument comes in. Are we explicitly requiring `valtype(metadata(x)) <: Tuple{Any,Symbol}`?

bkamins commented:
> Are we explicitly requiring `valtype(metadata(x)) <: Tuple{Any,Symbol}?`

Yes, that is why I assume that this is not lazy. In general this dictionary would have to be freshly allocated (actually both if `style=false` and if `style=true` just with different values). For this reason, if performance matters `metadatakeys` should be used as it is non-allocating

Tokazama commented:
That makes a lot of sense. Should we have some sort of documentation that makes it clear that `setindex!(metadata(x),...)` isn't the same as `metadata!(x...)`?

nalimilan commented on src/DataAPI.jl:
```diff
@@ -343,6 +343,9 @@ $STYLE_INFO
 If `default` is passed then return it if reading metadata is supported but
 mapping for `key` is missing. If `style=true` return `(default, :default)`.
+
+If `key` is not passed return a dictionary mapping all metadata keys to
+metadata values associated with object `x`.
```
I would document this as a separate method given that a lot of the text above doesn't apply when `key` isn't passed.

bkamins replied:
done

nalimilan commented:
Makes sense. I also think the docs should mention that the resulting object should not be modified as it is owned by `x`, and that it may have to be freshly allocated.

nalimilan commented:
Also, maybe we should define a fallback `Dict(key => metadata(df, key) for key in metadatakeys(df))`?

bkamins commented:
Following the comments I have:
1. implemented fallback methods for `metadata` and `colmetadata`
2. split docstings
3. added a comment both for `metadata` and `colmetadata` that the returned dictionary should be only used for reading data and that it may be freshly allocated.

nalimilan commented on src/DataAPI.jl:
```diff
@@ -343,12 +343,31 @@ $STYLE_INFO
+$STYLE_INFO
+
+The returned dictionary may be freshly allocated on each call to `metadata` and
+is considered to be owned by `x` so it should be only used for reading data.
```
```suggestion
is considered to be owned by `x` so it must not be modified.
```

nalimilan commented on src/DataAPI.jl:
```diff
@@ -404,6 +426,42 @@ If `style=true` return `(default, :default)`.
+$STYLE_INFO
+
+If `col` is not passed return a dictionary mapping columns represented as
+`Symbol` that have associated metadata to dictionaries dictionary mapping all
```
```suggestion
`Symbol` that have associated metadata to dictionaries mapping all
```

nalimilan commented on src/DataAPI.jl:
```diff
@@ -404,6 +426,42 @@ If `style=true` return `(default, :default)`.
+metadata keys to metadata values associated with table `x` for a given column.
+
+The returned dictionary may be freshly allocated on each call to `colmetadata`
+and is considered to be owned by `x` so it should be only used for reading
+data.
```
```suggestion
and is considered to be owned by `x` so it must not be modified.
```

nalimilan commented on src/DataAPI.jl:
```diff
@@ -404,6 +426,42 @@ If `style=true` return `(default, :default)`.
+
+$STYLE_INFO
+
+If `col` is not passed return a dictionary mapping columns represented as
+`Symbol` that have associated metadata to dictionaries dictionary mapping all
```
Maybe put this in the first paragraph, as it completely changes the behavior of the function? (It could amost make sense to have a separate docstring but that would be quite heavy...)

bkamins replied:
OK - done.

bkamins commented:
@ExpandingMan - can you please have a look at this PR before it is merged and released (it will have a minor in fluence on Parquet2.jl implementation, but actually it will go along what you partially have already). Thank you!

nalimilan commented on src/DataAPI.jl:
```diff
@@ -429,7 +429,11 @@ function colmetadata end
+`Symbol` that have associated metadata to dictionaries mapping all
+metadata keys to metadata values associated with table `x` for a given column.
+
+If `col` is passed a dictionary mapping all column metadata keys to metadata values
```
```suggestion
If `col` is passed return a dictionary mapping all column metadata keys to metadata values
```

ExpandingMan commented:
Looks good to me.

I haven't actually tested anything against this but it looks non-breaking.

bkamins commented:
Thank you!
